lib/Backend/Vault/Kv.hs

```diff
@@ -0,0 +1,231 @@
+        baseUrlToText = Just . T.pack . showBaseUrl
+        textToBaseUrl = parseBaseUrl . T.unpack
+
+data FieldMetadata = FieldMetadata
+  { _dateModified :: UTCTime
+  , _private :: Bool
+  }
+  deriving stock (Show, Generic)
+  deriving anyclass (A.ToJSON, A.FromJSON)
+makeLenses ''FieldMetadata
+
+data CofferSpecials =
+  CofferSpecials
+  { _masterKey :: Maybe T.Text
+  , _globalDateModified :: UTCTime
+  , _fields :: HS.HashMap T.Text FieldMetadata
+  , _tags :: [T.Text]
+  }
+  deriving stock (Show, Generic)
+  deriving anyclass (A.ToJSON, A.FromJSON)
+makeLenses ''CofferSpecials
+
+runVaultIO :: Member (Embed IO) r
+           => Member (Error CofferError) r
+           => BaseUrl
+           -> I.VaultToken
+           -> T.Text
+           -> Sem (BackendEffect ': r) a
+           -> Sem r a
+runVaultIO url token mount = interpret $
+  \x -> do
+    env <-
+      case url of
+        (BaseUrl Http _ _ _) -> do
+          manager <- embed $ newManager defaultManagerSettings
+          pure $ mkClientEnv manager url
+        (BaseUrl Https _ _ _) -> do
+          manager <- embed $ newManager tlsManagerSettings
+          pure $ mkClientEnv manager url
+
+    case x of
+      WriteSecret entry -> do
+        let
+          cofferSpecials = CofferSpecials
+            { _masterKey =
+              entry ^. E.masterField <&> E.getFieldKey
+            , _globalDateModified = entry ^. E.dateModified
+            , _fields =
+              entry ^. E.fields & traverse %~
+                (\f ->
+                FieldMetadata
+                { _dateModified = f ^. E.dateModified
+                , _private = f ^. E.private
+                })
+              & HS.mapKeys E.getFieldKey
+            , _tags = map E.getEntryTag $ entry ^. E.tags
+            }
+          secret = I.PostSecret
+            { I._cas = Nothing
+            , I._pdata =
+                  HS.insert "#$coffer" (TL.toStrict . TL.decodeUtf8 $ A.encode cofferSpecials)
+                . HS.map (^. E.value)
+                . HS.mapKeys E.getFieldKey
+                $ entry ^. E.fields
+            }
+
+        response <- embedCatch (entry ^. E.path) (postSecret env (entry ^. E.path) secret)
+
+        case response ^. I.ddata . at ("version" :: T.Text) of
+          Just (A.Number i) -> maybe (throw MarshallingFailed) pure (S.toBoundedInteger i)
+          _ -> throw MarshallingFailed
+      ReadSecret path version ->
+        runMaybe $ embedCatchMaybe path (readSecret env path version)
+        >>= \(I.KvResponse _ _ _ _ (I.ReadSecret _data _ _ _ _ _) _ _ _) -> do
+          cofferSpecials :: CofferSpecials <-
+            maybeThrow (_data ^.at "#$coffer" >>= A.decodeStrict' . T.encodeUtf8)
+          let secrets = HS.toList $ foldr HS.delete _data ["#$coffer"]
+          let keyToField key = do
+               _modTime <- cofferSpecials ^. fields.at key <&> (^. dateModified)
+               _private <- cofferSpecials ^. fields.at key <&> (^. private)
```

Comment:
You're mixing lens-style with non-lens-style in the same expression.

This could alternatively be expressed as:
```suggestion
               _private <- cofferSpecials ^? fields . at key . _Just . private
```

Reply:
Done